src/Infrastructure/LeanCode.Kratos/KratosWebHookHandlerBase.cs

```diff
@@ -0,0 +1,106 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using LeanCode.Kratos.Model;
+using Microsoft.AspNetCore.Http;
+
+namespace LeanCode.Kratos;
+
+public sealed record class KratosWebHookHandlerConfig(string ApiKey);
+
+public abstract partial class KratosWebHookHandlerBase
+{
+    private readonly Serilog.ILogger logger = Serilog.Log.ForContext<KratosWebHookHandlerBase>();
+
+    private readonly KratosWebHookHandlerConfig config;
+
+    protected KratosWebHookHandlerBase(KratosWebHookHandlerConfig config)
+    {
+        this.config = config;
+    }
+
+    protected virtual string ApiKeyHeaderName => "X-Api-Key";
+
+    protected abstract Task HandleCoreAsync(HttpContext ctx);
+
+    [SuppressMessage(
+        "?",
+        "CA1031",
+        Justification = "Exception is logged, rethrowing it for ASP.NET to handle is unnecessary."
+    )]
+    public async Task HandleAsync(HttpContext ctx)
+    {
+        try
+        {
+            if (VerifyApiKey(ctx.Request.Headers[ApiKeyHeaderName].ToString(), config.ApiKey))
+            {
+                await HandleCoreAsync(ctx);
+            }
+            else
+            {
+                logger.Error("Invalid Api Key");
+                ctx.Response.StatusCode = 403;
+            }
+        }
+        catch (Exception e)
+        {
+            logger.Error(e, "Failed to process webhook");
+            ctx.Response.StatusCode = 500;
+        }
+    }
+
+    private static bool VerifyApiKey(string left, string right)
+    {
+        var leftBytes = MemoryMarshal.AsBytes(left.AsSpan());
+        var rightBytes = MemoryMarshal.AsBytes(right.AsSpan());
+
+        return CryptographicOperations.FixedTimeEquals(leftBytes, rightBytes);
+    }
+
+    public static Task WriteIdentityResponseAsync(HttpContext ctx, Identity identity)
```

Comment:
What is the intended usage here? Only the derived class should call it? If yes - let's make this a `protected` method.

Reply:
In theory, there wouldn't be anything wrong with someone rolling their own base class (or not using any base class at all) and calling this there. But the intent was indeed about calling from subclasses so I can make it `protected` with little value being lost in the process.

Reply:
But it makes testing them much more annoying :slightly_frowning_face: